File: src/main/scala/com/bytebreakstudios/tictactoe/scala/condition/WinCondition.scala
package com.bytebreakstudios.tictactoe.scala.condition

import com.bytebreakstudios.board.Condition
import com.bytebreakstudios.tictactoe.scala.State
import com.bytebreakstudios.tictactoe.scala.action.WinAction

object WinCondition extends Condition[State, WinAction](new WinAction) {

  override def check(prev: State, cur: State): Boolean = {
    if (!prev.finished && !cur.finished) {
      val winner = testForWinner(cur)
      if (winner != 0){
        action.setSide(winner)
        return true
      }
    }
    false
  }

  private def testForWinner(s:State):Int = Seq(s.topRow, s.midRow, s.botRow, s.leftCol, s.midCol, s.rightCol, s.leftDiag, s.rightDiag).find(i => i == -3 || i == 3).getOrElse(0)

  override def description(): String = ???

  override def serialize(): String = ???

  override def deserialize(s: String) = ???
}

File: src/main/java/com/bytebreakstudios/tictactoe/condition/WinCondition.java
package com.bytebreakstudios.tictactoe.condition;

import com.bytebreakstudios.tictactoe.State;
import com.bytebreakstudios.tictactoe.action.WinAction;
import com.bytebreakstudios.board.Condition;

public class WinCondition extends Condition<State, WinAction> {
    public WinCondition(){
        super(new WinAction());
    }

    @Override
    public boolean check(State previous, State current) {
        if (!previous.finished && !current.finished){
            int winner = winner(current);
            if (winner != 0){
                action.setSide(winner);
                return true;
            }
        }
        return false;
    }

    private int winner(State s){
        int[] checks = new int[]{
                s.topRow(),
                s.midRow(),
                s.botRow(),
                s.leftCol(),
                s.midCol(),
                s.rightCol(),
                s.leftDiag(),
                s.rightDiag()
        };
        for (int check : checks){
            if (check == -3 || check == 3) return check;
        }
        return 0;
    }

    @Override
    public String description() {
        return null;
    }

    @Override
    public String serialize() {
        return null;
    }

    @Override
    public Condition deserialize(String s) {
        return null;
    }
}
